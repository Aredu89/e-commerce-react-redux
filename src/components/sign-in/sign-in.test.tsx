import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import SignIn from './sign-in.component';
import { googleSignInStart, emailSignInStart } from '../../store/user/user.action';

const mockedDispatch = jest.fn();
const mockedSetState = jest.fn();
const mockedUseState = ['', mockedSetState];

jest.mock('react', () => ({
  ...jest.requireActual('react'),
  useState: () => mockedUseState
}));

jest.mock('react-redux', () => ({
  ...jest.requireActual('react-redux'),
  useDispatch: () => mockedDispatch
}));

describe('Sign-in component', () => {
  it('renders properly', () => {
    const result = render(<SignIn />);
    expect(result).toMatchSnapshot();
  });
  it('captures email input', () => {
    render(<SignIn />);
    
    const input = screen.getByTestId('email-input');

    userEvent.type(input, 'hello');

    expect(mockedSetState).toHaveBeenCalledTimes(5);
    expect(mockedSetState).toHaveBeenCalledWith('h');
    expect(mockedSetState).toHaveBeenCalledWith('e');
    expect(mockedSetState).toHaveBeenCalledWith('l');
    expect(mockedSetState).toHaveBeenCalledWith('l');
    expect(mockedSetState).toHaveBeenCalledWith('o');
  });
  it('captures password input', () => {
    render(<SignIn />);
    
    const input = screen.getByTestId('password-input');

    userEvent.type(input, 'pass');
    expect(mockedSetState).toHaveBeenCalledTimes(4);
    expect(mockedSetState).toHaveBeenCalledWith('p');
    expect(mockedSetState).toHaveBeenCalledWith('a');
    expect(mockedSetState).toHaveBeenCalledWith('s');
    expect(mockedSetState).toHaveBeenCalledWith('s');
  });
  it('logs google user', () => {
    render(<SignIn />);
    const button = screen.getByText('SIGN IN WITH GOOGLE');

    userEvent.click(button);
    expect(mockedDispatch).toHaveBeenCalledWith(googleSignInStart());
  });
  it('submits', async () => {
    render(<SignIn />);
    const submitButton = screen.getByRole('button', { name: 'Sign in' });

    userEvent.click(submitButton);

    expect(mockedDispatch).toHaveBeenCalledWith(emailSignInStart('', ''));
  });
});
