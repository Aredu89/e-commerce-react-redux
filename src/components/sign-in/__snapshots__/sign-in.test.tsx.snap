// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sign-in component renders properly 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="sc-papXJ exRLuz"
      >
        <h2>
          I already have an account
        </h2>
        <span>
          Sign in with your email and password
        </span>
        <form>
          <div
            class="sc-ftvSup cduAcl"
          >
            <input
              class="sc-iBkjds eZAXkO"
              data-testid="email-input"
              name="email"
              required=""
              type="email"
              value=""
            />
            <label
              class="sc-gKXOVf gjofEz"
            >
              Email
            </label>
          </div>
          <div
            class="sc-ftvSup cduAcl"
          >
            <input
              class="sc-iBkjds eZAXkO"
              data-testid="password-input"
              name="password"
              required=""
              type="password"
              value=""
            />
            <label
              class="sc-gKXOVf gjofEz"
            >
              Password
            </label>
          </div>
          <div
            class="sc-jqUVSM eVqAl"
          >
            <button
              class="sc-dkzDqf hdlUww"
              name="Submit Form"
              type="submit"
              value="Submit Form"
            >
              Sign in
            </button>
            <button
              class="sc-dkzDqf sc-hKMtZM hdlUww fbHvTo"
              type="button"
            >
               
              SIGN IN WITH GOOGLE
               
            </button>
          </div>
        </form>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="sc-papXJ exRLuz"
    >
      <h2>
        I already have an account
      </h2>
      <span>
        Sign in with your email and password
      </span>
      <form>
        <div
          class="sc-ftvSup cduAcl"
        >
          <input
            class="sc-iBkjds eZAXkO"
            data-testid="email-input"
            name="email"
            required=""
            type="email"
            value=""
          />
          <label
            class="sc-gKXOVf gjofEz"
          >
            Email
          </label>
        </div>
        <div
          class="sc-ftvSup cduAcl"
        >
          <input
            class="sc-iBkjds eZAXkO"
            data-testid="password-input"
            name="password"
            required=""
            type="password"
            value=""
          />
          <label
            class="sc-gKXOVf gjofEz"
          >
            Password
          </label>
        </div>
        <div
          class="sc-jqUVSM eVqAl"
        >
          <button
            class="sc-dkzDqf hdlUww"
            name="Submit Form"
            type="submit"
            value="Submit Form"
          >
            Sign in
          </button>
          <button
            class="sc-dkzDqf sc-hKMtZM hdlUww fbHvTo"
            type="button"
          >
             
            SIGN IN WITH GOOGLE
             
          </button>
        </div>
      </form>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
